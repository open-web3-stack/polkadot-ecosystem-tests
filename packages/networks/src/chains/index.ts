export * from './acala.js'
export * from './assethub.js'
export * from './astar.js'
export * from './hydraDX.js'
export * from './moonbeam.js'
export * from './polkadot.js'
