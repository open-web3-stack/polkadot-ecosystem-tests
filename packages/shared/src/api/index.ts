import type { ApiPromise } from '@polkadot/api'

export type TransferType = 'Teleport' | 'LocalReserve' | 'DestinationReserve' | 'RemoteReserve'

export const xtokens = {
  relaychainV3: (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            id: acc,
          },
        },
      },
    },
  }),
  relaychainV4: (acc: any) => ({
    V4: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            id: acc,
          },
        },
      },
    },
  }),
  parachainAccountId20V3: (paraId: number) => (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountKey20: {
              key: acc,
            },
          },
        ],
      },
    },
  }),
  parachainV3: (paraId: number) => (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountId32: {
              id: acc,
            },
          },
        ],
      },
    },
  }),
  parachainV4: (paraId: number) => (acc: any) => ({
    V4: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountId32: {
              id: acc,
            },
          },
        ],
      },
    },
  }),
  transfer:
    (token: any, amount: any, dest: (dest: any) => any, weight: any = 'Unlimited') =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      api.tx.xTokens.transfer(token, amount, dest(acc), weight),
  transferMulticurrencies:
    (token: any, amount: any, feeToken: any, feeAmount: any, dest: (dest: any) => any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      api.tx.xTokens.transferMulticurrencies(
        [
          [token, amount],
          [feeToken, feeAmount],
        ],
        1,
        dest(acc),
        'Unlimited',
      ),
}

export const xcmPallet = {
  relaychainV4: {
    V4: {
      parents: 1,
      interior: 'Here',
    },
  },
  parachainV3: (parents: number, paraId: any) => ({
    V3: {
      parents,
      interior: {
        X1: { Parachain: paraId },
      },
    },
  }),
  parachainV4: (parents: number, paraId: any) => ({
    V4: {
      parents,
      interior: {
        X1: [{ Parachain: paraId }],
      },
    },
  }),
  limitedTeleportAssets:
    (token: any, amount: any, dest: any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).limitedTeleportAssets(
        dest,
        {
          V3: {
            parents: 0,
            interior: {
              X1: {
                AccountId32: {
                  // network: 'Any',
                  id: acc,
                },
              },
            },
          },
        },
        {
          V3: [
            {
              id: token,
              fun: { Fungible: amount },
            },
          ],
        },
        0,
        'Unlimited',
      ),
  limitedReserveTransferAssetsV3:
    (token: any, amount: any, dest: any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
        dest,
        {
          V3: {
            parents: 0,
            interior: {
              X1: {
                AccountId32: {
                  id: acc,
                },
              },
            },
          },
        },
        {
          V3: [
            {
              id: token,
              fun: { Fungible: amount },
            },
          ],
        },
        0,
        'Unlimited',
      ),
  teleportAssetsV3:
    (token: any, amount: any, dest: any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).teleportAssets(
        dest,
        {
          V3: {
            parents: 0,
            interior: {
              X1: {
                AccountId32: {
                  id: acc,
                },
              },
            },
          },
        },
        {
          V3: [
            {
              id: token,
              fun: { Fungible: amount },
            },
          ],
        },
        0,
      ),
  executeXCM:
    (xcm: any, max_weight: any) =>
    ({ api }: { api: ApiPromise }) => {
      return (api.tx.xcmPallet || api.tx.polkadotXcm).execute(xcm, max_weight)
    },
  transferAssetsV3:
    (token: any, amount: any, dest: any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).transferAssets(
        dest,
        {
          V3: {
            parents: 0,
            interior: {
              X1: {
                AccountId32: {
                  id: acc,
                },
              },
            },
          },
        },
        {
          V3: [
            {
              id: token,
              fun: { Fungible: amount },
            },
          ],
        },
        0,
        'Unlimited',
      ),
  transferAssetsUsingType:
    (dest: any, tokens: any[], assetTransferType: TransferType, remoteFeesId: any, feesTransferType: TransferType) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).transferAssetsUsingTypeAndThen(
        dest,
        { V4: tokens },
        assetTransferType,
        { V4: remoteFeesId },
        feesTransferType,
        {
          V4: [
            {
              DepositAsset: {
                assets: {
                  Wild: {
                    AllCounted: 2,
                  },
                },
                beneficiary: {
                  parents: 0,
                  interior: { X1: [{ AccountId32: { id: acc } }] },
                },
              },
            },
          ],
        },
        'Unlimited',
      ),
}

export const tx = {
  xtokens,
  xcmPallet,
}

export const query = {
  balances: ({ api }: { api: ApiPromise }, address: string) => api.query.system.account(address),
  totalIssuance: ({ api }: { api: ApiPromise }) => api.query.balances.totalIssuance(),
  tokens:
    (token: any) =>
    ({ api }: { api: ApiPromise }, address: string) =>
      api.query.tokens.accounts(address, token),
  assets:
    (token: number | bigint) =>
    ({ api }: { api: ApiPromise }, address: string) =>
      api.query.assets.account(token, address),
  foreignAssets:
    (token: any) =>
    ({ api }: { api: ApiPromise }, address: string) =>
      api.query.foreignAssets.account(token, address),
  evm:
    (contract: string, slot: string) =>
    ({ api }: { api: ApiPromise }, _address: string) =>
      api.query.evm.accountStorages(contract, slot),
}
