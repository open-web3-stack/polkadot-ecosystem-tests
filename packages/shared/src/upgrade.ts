import { sendTransaction } from '@acala-network/chopsticks-testing'

import { type Chain, type Client, defaultAccounts } from '@e2e-test/networks'
import { sendWhitelistCallViaXcmTransact } from '@e2e-test/shared'

import type { HexString } from '@polkadot/util/types'

import { assert } from 'vitest'

import { checkEvents, checkSystemEvents, createXcmTransactSend, scheduleInlineCallWithOrigin } from './helpers/index.js'

/**
 * Computes the XCM `MultiLocation` route from a source chain to a destination chain.
 *
 * @param from - The source chain (the chain initiating the XCM message).
 * @param to - The destination chain (the chain intended to receive and execute the XCM message).
 */
function getXcmRoute(from: Chain, to: Chain) {
  let parents: number
  let interior: any

  if (from.isRelayChain) {
    parents = 0
  } else {
    parents = 1
  }

  if (to.isRelayChain) {
    interior = 'Here'
  } else {
    interior = { X1: [{ Parachain: to.paraId }] }
  }

  return { parents, interior }
}

/**
 * Constructs an XCM forceBatch transaction that authorizes a runtime upgrade on a destination chain.
 *
 *
 * @param codeHash - The code hash of the new runtime to be authorized.
 * @param sourceClient - The client instance representing the source chain (where the XCM is sent from).
 * @param destClient - The client instance representing the destination chain (where the upgrade is authorized).
 */
export function createXcmAuthorizeUpgradeBatch(codeHash: HexString, sourceClient: Client, destClient: Client) {
  const authorizeUpgradeCall = destClient.api.tx.system.authorizeUpgrade(codeHash)
  const dest = getXcmRoute(sourceClient.config, destClient.config)

  const xcmTx = createXcmTransactSend(sourceClient, dest, authorizeUpgradeCall.method.toHex(), 'Superuser', {
    refTime: '5000000000',
    proofSize: '500000',
  })

  return sourceClient.api.tx.utility.forceBatch([xcmTx])
}

/**
 * Simulates and executes the full governance workflow to authorize a runtime upgrade
 * on a target chain, optionally across chains using collectives and XCM.
 *
 * - Considering any existing authorized upgrade
 * - Submitting the upgrade call preimage
 * - Attempt to dispatch when call is not yet whitelisted
 * - Whitelisting the call via a Collectives chain
 * - Attempting to dispatch with both valid and invalid origins.
 * - Validating that the upgrade was correctly authorized on the target chain.
 *
 * @param governingChain - The chain where the governance is running, allowed to execute as superuser on other chains
 * @param chainToUpgrade - The chain whose runtime is being upgraded.
 * @param collectivesChain - The chain that hosts the collective body which will whitelist the upgrade call.
 */
export async function authorizeUpgradeViaCollectives(
  governingChain: Client,
  chainToUpgrade: Client,
  collectivesChain: Client,
) {
  // NOTE: Since the test is run against some live chain data, it may happen that at some moment some upgrade
  //       is already authorized - expected result is that the authorized hash will be overriden by this test
  //
  // specific value of codeHash is not important for the test
  const codeHash = '0x0101010101010101010101010101010101010101010101010101010101010101'
  const assertAuthorizedUpgradeUnchanged = async () => {
    const currentAuthorizedUpgrade = await chainToUpgrade.api.query.system.authorizedUpgrade()
    assert(currentAuthorizedUpgrade.isNone || currentAuthorizedUpgrade.value.codeHash.toHex() !== codeHash)
  }
  await assertAuthorizedUpgradeUnchanged()

  const authorizeUpgradeCall =
    governingChain.url === chainToUpgrade.url
      ? chainToUpgrade.api.tx.system.authorizeUpgrade(codeHash)
      : createXcmAuthorizeUpgradeBatch(codeHash, governingChain, chainToUpgrade)

  const whiteListCall = governingChain.api.tx.whitelist.dispatchWhitelistedCallWithPreimage(
    authorizeUpgradeCall.method.toHex(),
  )

  const notePreimageTx = governingChain.api.tx.preimage.notePreimage(authorizeUpgradeCall.method.toHex())

  const okOrigin = { Origins: 'WhitelistedCaller' }
  // some existing but not allowed to dispatch whitelisted call origin
  const badOrigin = { Origins: 'StakingAdmin' }

  // it does not matter who is the origin which notes the preimage
  const noteEvents = await sendTransaction(notePreimageTx.signAsync(defaultAccounts.alice))
  await governingChain.dev.newBlock()
  await checkEvents(noteEvents, 'preimage').toMatchSnapshot('events after notePreimge')

  // try to dispatch a call that has not yet been whitelisted - should result in dispatching error
  await scheduleInlineCallWithOrigin(governingChain, whiteListCall.method.toHex(), okOrigin)
  await governingChain.dev.newBlock()
  await checkSystemEvents(governingChain, 'scheduler')
    .redact({ hash: false, redactKeys: /task/ })
    .toMatchSnapshot('events when dispatching non-whitelisted call')
  await assertAuthorizedUpgradeUnchanged()

  // collectives whitelisting a call
  await sendWhitelistCallViaXcmTransact(governingChain, collectivesChain, authorizeUpgradeCall.method.hash.toHex(), {
    proofSize: '10000',
    refTime: '500000000',
  })
  await collectivesChain.dev.newBlock()
  await checkSystemEvents(collectivesChain, 'polkadotXcm')
    .redact({ hash: false, redactKeys: /messageId/ })
    .toMatchSnapshot('collectives events emitted when sending xcm')
  await governingChain.dev.newBlock()
  await checkSystemEvents(governingChain, 'whitelist', 'messageQueue')
    .redact({ hash: false, redactKeys: /id/ })
    .toMatchSnapshot('governing chain events emitted on receiving xcm from collectives')

  // trying to dispatch whitelisted call using bad origin - should result in error
  await scheduleInlineCallWithOrigin(governingChain, whiteListCall.method.toHex(), badOrigin)
  await governingChain.dev.newBlock()
  await checkSystemEvents(governingChain, 'scheduler')
    .redact({ hash: false, redactKeys: /task/ })
    .toMatchSnapshot('events when dispatching whitelisted call with bad origin')
  await assertAuthorizedUpgradeUnchanged()

  // call is whitelisted, origin is ok - success expected
  await scheduleInlineCallWithOrigin(governingChain, whiteListCall.method.toHex(), okOrigin)
  await governingChain.dev.newBlock()
  await checkSystemEvents(governingChain, 'whitelist')
    .redact({ hash: false })
    .toMatchSnapshot('governing chain events about dispatching whitelisted call')

  if (governingChain.url === chainToUpgrade.url) {
    await checkSystemEvents(chainToUpgrade, { section: 'system', method: 'UpgradeAuthorized' })
      .redact({ hash: false })
      .toMatchSnapshot('to-be-upgraded chain events to confirm authorized upgrade')
  } else {
    await chainToUpgrade.dev.newBlock()
    await checkSystemEvents(chainToUpgrade, 'messageQueue', { section: 'system', method: 'UpgradeAuthorized' })
      .redact({ hash: false, redactKeys: /id/ })
      .toMatchSnapshot('to-be-upgraded chain events to confirm authorized upgrade')
  }

  assert.equal((await chainToUpgrade.api.query.system.authorizedUpgrade()).value.codeHash.toHex(), codeHash)
}
