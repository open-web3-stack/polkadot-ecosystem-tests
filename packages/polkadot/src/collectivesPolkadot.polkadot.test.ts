import { collectivesPolkadot, polkadot } from '@e2e-test/networks/chains'
import { collectivesChainE2ETests, setupNetworks } from '@e2e-test/shared'
import { query, tx } from '@e2e-test/shared/api'
import { authorizeUpgradeViaCollectives } from '@e2e-test/shared/upgrade.js'
import { runXcmPalletDown, runXcmPalletUp } from '@e2e-test/shared/xcm'

import { describe, test } from 'vitest'

collectivesChainE2ETests(polkadot, collectivesPolkadot, { testSuiteName: 'collectives & polkadot' })

describe('collectives & polkadot', async () => {
  const [polkadotClient, collectivesClient] = await setupNetworks(polkadot, collectivesPolkadot)

  const collectivesDOT = collectivesPolkadot.custom.dot
  const polkadotDOT = polkadot.custom.dot

  runXcmPalletDown('polkadot teleport DOT to collectivesPolkadot', async () => {
    return {
      fromChain: polkadotClient,
      toChain: collectivesClient,
      balance: query.balances,
      tx: tx.xcmPallet.teleportAssetsV3(polkadotDOT, 1e12, tx.xcmPallet.parachainV3(0, collectivesPolkadot.paraId!)),
      totalIssuanceProvider: () => query.totalIssuance(polkadotClient),
    }
  })

  runXcmPalletUp('collectivesPolkadot teleport DOT to polkadot', async () => {
    return {
      fromChain: collectivesClient,
      toChain: polkadotClient,
      balance: query.balances,
      tx: tx.xcmPallet.teleportAssetsV3(collectivesDOT, 1e12, tx.xcmPallet.relaychainV4),
      totalIssuanceProvider: () => query.totalIssuance(polkadotClient),
    }
  })

  test('Relay authorizes upgrade for itself', async () => {
    await authorizeUpgradeViaCollectives(polkadotClient, polkadotClient, collectivesClient)
  })

  test('Relay authorizes Collectives upgrade via Collectives', async () => {
    await authorizeUpgradeViaCollectives(polkadotClient, collectivesClient, collectivesClient)
  })
})
