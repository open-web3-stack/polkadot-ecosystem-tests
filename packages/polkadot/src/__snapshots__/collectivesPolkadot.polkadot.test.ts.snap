// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`collectives & polkadot > Relay authorizes Collectives upgrade via Collectives > collectives events emitted when sending xcm 1`] = `
[
  {
    "data": {
      "fees": [
        {
          "fun": {
            "Fungible": "(rounded 300000000)",
          },
          "id": {
            "interior": "Here",
            "parents": 1,
          },
        },
      ],
      "paying": {
        "interior": {
          "X1": [
            {
              "Plurality": {
                "id": "Technical",
                "part": "Voice",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "FeesPaid",
    "section": "polkadotXcm",
  },
  {
    "data": {
      "destination": {
        "interior": "Here",
        "parents": 1,
      },
      "message": [
        {
          "UnpaidExecution": {
            "checkOrigin": null,
            "weightLimit": "Unlimited",
          },
        },
        {
          "Transact": {
            "call": {
              "encoded": "0x17001f316948a52772e24450da99383b0e57b75770f95e916e945b6fd91163ce05dd",
            },
            "originKind": "Xcm",
            "requireWeightAtMost": {
              "proofSize": 10000,
              "refTime": 500000000,
            },
          },
        },
      ],
      "messageId": "(redacted)",
      "origin": {
        "interior": {
          "X1": [
            {
              "Plurality": {
                "id": "Technical",
                "part": "Voice",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "Sent",
    "section": "polkadotXcm",
  },
]
`;

exports[`collectives & polkadot > Relay authorizes Collectives upgrade via Collectives > events after notePreimge 1`] = `
[
  {
    "data": {
      "hash_": "(hash)",
    },
    "method": "Noted",
    "section": "preimage",
  },
]
`;

exports[`collectives & polkadot > Relay authorizes Collectives upgrade via Collectives > events when dispatching non-whitelisted call 1`] = `
[
  {
    "data": {
      "id": null,
      "result": {
        "Err": {
          "Module": {
            "error": "0x03000000",
            "index": 23,
          },
        },
      },
      "task": "(redacted)",
    },
    "method": "Dispatched",
    "section": "scheduler",
  },
]
`;

exports[`collectives & polkadot > Relay authorizes Collectives upgrade via Collectives > events when dispatching whitelisted call with bad origin 1`] = `
[
  {
    "data": {
      "id": null,
      "result": {
        "Err": "BadOrigin",
      },
      "task": "(redacted)",
    },
    "method": "Dispatched",
    "section": "scheduler",
  },
]
`;

exports[`collectives & polkadot > Relay authorizes Collectives upgrade via Collectives > governing chain events about dispatching whitelisted call 1`] = `
[
  {
    "data": {
      "callHash": "0x1f316948a52772e24450da99383b0e57b75770f95e916e945b6fd91163ce05dd",
      "result": {
        "Ok": {
          "actualWeight": {
            "proofSize": "(rounded 3600)",
            "refTime": "(rounded 300000000)",
          },
          "paysFee": "Yes",
        },
      },
    },
    "method": "WhitelistedCallDispatched",
    "section": "whitelist",
  },
]
`;

exports[`collectives & polkadot > Relay authorizes Collectives upgrade via Collectives > governing chain events emitted on receiving xcm from collectives 1`] = `
[
  {
    "data": {
      "callHash": "0x1f316948a52772e24450da99383b0e57b75770f95e916e945b6fd91163ce05dd",
    },
    "method": "CallWhitelisted",
    "section": "whitelist",
  },
  {
    "data": {
      "id": "(redacted)",
      "origin": {
        "Ump": {
          "Para": "(rounded 1000)",
        },
      },
      "success": true,
      "weightUsed": {
        "proofSize": 10000,
        "refTime": "(rounded 510000000)",
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
]
`;

exports[`collectives & polkadot > Relay authorizes Collectives upgrade via Collectives > to-be-upgraded chain events to confirm authorized upgrade 1`] = `
[
  {
    "data": {
      "checkVersion": true,
      "codeHash": "0x0101010101010101010101010101010101010101010101010101010101010101",
    },
    "method": "UpgradeAuthorized",
    "section": "system",
  },
  {
    "data": {
      "id": "(redacted)",
      "origin": "Parent",
      "success": true,
      "weightUsed": {
        "proofSize": 500000,
        "refTime": "(rounded 5000000000)",
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
]
`;

exports[`collectives & polkadot > Relay authorizes upgrade for itself > collectives events emitted when sending xcm 1`] = `
[
  {
    "data": {
      "fees": [
        {
          "fun": {
            "Fungible": "(rounded 300000000)",
          },
          "id": {
            "interior": "Here",
            "parents": 1,
          },
        },
      ],
      "paying": {
        "interior": {
          "X1": [
            {
              "Plurality": {
                "id": "Technical",
                "part": "Voice",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "FeesPaid",
    "section": "polkadotXcm",
  },
  {
    "data": {
      "destination": {
        "interior": "Here",
        "parents": 1,
      },
      "message": [
        {
          "UnpaidExecution": {
            "checkOrigin": null,
            "weightLimit": "Unlimited",
          },
        },
        {
          "Transact": {
            "call": {
              "encoded": "0x17003812e839c2f2db66069c9ee359bbc78e7ec08cb657d0a735742e68c5c0668a84",
            },
            "originKind": "Xcm",
            "requireWeightAtMost": {
              "proofSize": 10000,
              "refTime": 500000000,
            },
          },
        },
      ],
      "messageId": "(redacted)",
      "origin": {
        "interior": {
          "X1": [
            {
              "Plurality": {
                "id": "Technical",
                "part": "Voice",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "Sent",
    "section": "polkadotXcm",
  },
]
`;

exports[`collectives & polkadot > Relay authorizes upgrade for itself > events after notePreimge 1`] = `
[
  {
    "data": {
      "hash_": "(hash)",
    },
    "method": "Noted",
    "section": "preimage",
  },
]
`;

exports[`collectives & polkadot > Relay authorizes upgrade for itself > events when dispatching non-whitelisted call 1`] = `
[
  {
    "data": {
      "id": null,
      "result": {
        "Err": {
          "Module": {
            "error": "0x03000000",
            "index": 23,
          },
        },
      },
      "task": "(redacted)",
    },
    "method": "Dispatched",
    "section": "scheduler",
  },
]
`;

exports[`collectives & polkadot > Relay authorizes upgrade for itself > events when dispatching whitelisted call with bad origin 1`] = `
[
  {
    "data": {
      "id": null,
      "result": {
        "Err": "BadOrigin",
      },
      "task": "(redacted)",
    },
    "method": "Dispatched",
    "section": "scheduler",
  },
]
`;

exports[`collectives & polkadot > Relay authorizes upgrade for itself > governing chain events about dispatching whitelisted call 1`] = `
[
  {
    "data": {
      "callHash": "0x3812e839c2f2db66069c9ee359bbc78e7ec08cb657d0a735742e68c5c0668a84",
      "result": {
        "Ok": {
          "actualWeight": null,
          "paysFee": "Yes",
        },
      },
    },
    "method": "WhitelistedCallDispatched",
    "section": "whitelist",
  },
]
`;

exports[`collectives & polkadot > Relay authorizes upgrade for itself > governing chain events emitted on receiving xcm from collectives 1`] = `
[
  {
    "data": {
      "callHash": "0x3812e839c2f2db66069c9ee359bbc78e7ec08cb657d0a735742e68c5c0668a84",
    },
    "method": "CallWhitelisted",
    "section": "whitelist",
  },
  {
    "data": {
      "id": "(redacted)",
      "origin": {
        "Ump": {
          "Para": "(rounded 1000)",
        },
      },
      "success": true,
      "weightUsed": {
        "proofSize": 10000,
        "refTime": "(rounded 510000000)",
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
]
`;

exports[`collectives & polkadot > Relay authorizes upgrade for itself > to-be-upgraded chain events to confirm authorized upgrade 1`] = `
[
  {
    "data": {
      "checkVersion": true,
      "codeHash": "0x0101010101010101010101010101010101010101010101010101010101010101",
    },
    "method": "UpgradeAuthorized",
    "section": "system",
  },
]
`;

exports[`collectives & polkadot > collectivesPolkadot teleport DOT to polkadot > balance on from chain 1`] = `
{
  "consumers": 0,
  "data": {
    "flags": "0x80000000000000000000000000000000",
    "free": "(rounded 9000000000000)",
    "frozen": 0,
    "reserved": 0,
  },
  "nonce": 1,
  "providers": 1,
  "sufficients": 0,
}
`;

exports[`collectives & polkadot > collectivesPolkadot teleport DOT to polkadot > balance on to chain 1`] = `
{
  "consumers": 0,
  "data": {
    "flags": "0x80000000000000000000000000000000",
    "free": "(rounded 1000000000000)",
    "frozen": 0,
    "reserved": 0,
  },
  "nonce": 0,
  "providers": 1,
  "sufficients": 0,
}
`;

exports[`collectives & polkadot > collectivesPolkadot teleport DOT to polkadot > from chain ump messages 1`] = `
[
  {
    "v4": [
      {
        "receiveTeleportedAsset": [
          {
            "fun": {
              "fungible": 1000000000000,
            },
            "id": {
              "interior": {
                "here": null,
              },
              "parents": 0,
            },
          },
        ],
      },
      {
        "clearOrigin": null,
      },
      {
        "buyExecution": {
          "fees": {
            "fun": {
              "fungible": 1000000000000,
            },
            "id": {
              "interior": {
                "here": null,
              },
              "parents": 0,
            },
          },
          "weightLimit": {
            "unlimited": null,
          },
        },
      },
      {
        "depositAsset": {
          "assets": {
            "wild": {
              "allCounted": 1,
            },
          },
          "beneficiary": {
            "interior": {
              "x1": [
                {
                  "accountId32": {
                    "id": "0xd17c2d7823ebf260fd138f2d7e27d114c0145d968b5ff5006125f2414fadae69",
                    "network": null,
                  },
                },
              ],
            },
            "parents": 0,
          },
        },
      },
      {
        "setTopic": "(redacted)",
      },
    ],
  },
]
`;

exports[`collectives & polkadot > collectivesPolkadot teleport DOT to polkadot > to chain ump events 1`] = `
[
  {
    "data": {
      "id": "(hash)",
      "origin": {
        "Ump": {
          "Para": "(rounded 1000)",
        },
      },
      "success": true,
      "weightUsed": {
        "proofSize": "(rounded 7200)",
        "refTime": "(rounded 280000000)",
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
]
`;

exports[`collectives & polkadot > collectivesPolkadot teleport DOT to polkadot > tx events 1`] = `
[
  {
    "data": {
      "outcome": {
        "Complete": {
          "used": {
            "proofSize": "(rounded 3590)",
            "refTime": "(rounded 168000000)",
          },
        },
      },
    },
    "method": "Attempted",
    "section": "polkadotXcm",
  },
  {
    "data": {
      "fees": [
        {
          "fun": {
            "Fungible": "(rounded 305000000)",
          },
          "id": {
            "interior": "Here",
            "parents": 1,
          },
        },
      ],
      "paying": {
        "interior": {
          "X1": [
            {
              "AccountId32": {
                "id": "(hash)",
                "network": "Polkadot",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "FeesPaid",
    "section": "polkadotXcm",
  },
  {
    "data": {
      "destination": {
        "interior": "Here",
        "parents": 1,
      },
      "message": [
        {
          "ReceiveTeleportedAsset": [
            {
              "fun": {
                "Fungible": 1000000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 0,
              },
            },
          ],
        },
        "ClearOrigin",
        {
          "BuyExecution": {
            "fees": {
              "fun": {
                "Fungible": 1000000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 0,
              },
            },
            "weightLimit": "Unlimited",
          },
        },
        {
          "DepositAsset": {
            "assets": {
              "Wild": {
                "AllCounted": 1,
              },
            },
            "beneficiary": {
              "interior": {
                "X1": [
                  {
                    "AccountId32": {
                      "id": "(hash)",
                      "network": null,
                    },
                  },
                ],
              },
              "parents": 0,
            },
          },
        },
      ],
      "messageId": "(hash)",
      "origin": {
        "interior": {
          "X1": [
            {
              "AccountId32": {
                "id": "(hash)",
                "network": "Polkadot",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "Sent",
    "section": "polkadotXcm",
  },
]
`;

exports[`collectives & polkadot > polkadot teleport DOT to collectivesPolkadot > balance on from chain 1`] = `
{
  "consumers": 0,
  "data": {
    "flags": "0x80000000000000000000000000000000",
    "free": "(rounded 9000000000000)",
    "frozen": 0,
    "reserved": 0,
  },
  "nonce": 1,
  "providers": 1,
  "sufficients": 0,
}
`;

exports[`collectives & polkadot > polkadot teleport DOT to collectivesPolkadot > balance on to chain 1`] = `
{
  "consumers": 0,
  "data": {
    "flags": "0x80000000000000000000000000000000",
    "free": "(rounded 11000000000000)",
    "frozen": 0,
    "reserved": 0,
  },
  "nonce": 0,
  "providers": 1,
  "sufficients": 0,
}
`;

exports[`collectives & polkadot > polkadot teleport DOT to collectivesPolkadot > to chain dmp events 1`] = `
[
  {
    "data": {
      "count": 1,
    },
    "method": "DownwardMessagesReceived",
    "section": "parachainSystem",
  },
  {
    "data": {
      "dmqHead": "(hash)",
      "weightUsed": {
        "proofSize": "(rounded 3500)",
        "refTime": "(rounded 760000000)",
      },
    },
    "method": "DownwardMessagesProcessed",
    "section": "parachainSystem",
  },
  {
    "data": {
      "id": "(hash)",
      "origin": "Parent",
      "success": true,
      "weightUsed": {
        "proofSize": "(rounded 3600)",
        "refTime": "(rounded 160000000)",
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
]
`;

exports[`collectives & polkadot > polkadot teleport DOT to collectivesPolkadot > tx events 1`] = `
[
  {
    "data": {
      "outcome": {
        "Complete": {
          "used": {
            "proofSize": "(rounded 3590)",
            "refTime": "(rounded 144000000)",
          },
        },
      },
    },
    "method": "Attempted",
    "section": "xcmPallet",
  },
  {
    "data": {
      "fees": [
        {
          "fun": {
            "Fungible": 397000000,
          },
          "id": {
            "interior": "Here",
            "parents": 0,
          },
        },
      ],
      "paying": {
        "interior": {
          "X1": [
            {
              "AccountId32": {
                "id": "(hash)",
                "network": "Polkadot",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "FeesPaid",
    "section": "xcmPallet",
  },
  {
    "data": {
      "destination": {
        "interior": {
          "X1": [
            {
              "Parachain": "(rounded 1000)",
            },
          ],
        },
        "parents": 0,
      },
      "message": [
        {
          "ReceiveTeleportedAsset": [
            {
              "fun": {
                "Fungible": 1000000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 1,
              },
            },
          ],
        },
        "ClearOrigin",
        {
          "BuyExecution": {
            "fees": {
              "fun": {
                "Fungible": 1000000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 1,
              },
            },
            "weightLimit": "Unlimited",
          },
        },
        {
          "DepositAsset": {
            "assets": {
              "Wild": {
                "AllCounted": 1,
              },
            },
            "beneficiary": {
              "interior": {
                "X1": [
                  {
                    "AccountId32": {
                      "id": "(hash)",
                      "network": null,
                    },
                  },
                ],
              },
              "parents": 0,
            },
          },
        },
      ],
      "messageId": "(hash)",
      "origin": {
        "interior": {
          "X1": [
            {
              "AccountId32": {
                "id": "(hash)",
                "network": "Polkadot",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "Sent",
    "section": "xcmPallet",
  },
]
`;

exports[`collectives & polkadot > whitelisting a call by fellowship > destination chain events 1`] = `
[
  {
    "data": {
      "callHash": "0x0101010101010101010101010101010101010101010101010101010101010101",
    },
    "method": "CallWhitelisted",
    "section": "whitelist",
  },
  {
    "data": {
      "id": "(redacted)",
      "origin": {
        "Ump": {
          "Para": "(rounded 1000)",
        },
      },
      "success": true,
      "weightUsed": {
        "proofSize": 10000,
        "refTime": "(rounded 1000000000)",
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
]
`;

exports[`collectives & polkadot > whitelisting a call by fellowship > source chain events 1`] = `
[
  {
    "data": {
      "fees": [
        {
          "fun": {
            "Fungible": "(rounded 300000000)",
          },
          "id": {
            "interior": "Here",
            "parents": 1,
          },
        },
      ],
      "paying": {
        "interior": {
          "X1": [
            {
              "Plurality": {
                "id": "Technical",
                "part": "Voice",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "FeesPaid",
    "section": "polkadotXcm",
  },
  {
    "data": {
      "destination": {
        "interior": "Here",
        "parents": 1,
      },
      "message": [
        {
          "UnpaidExecution": {
            "checkOrigin": null,
            "weightLimit": "Unlimited",
          },
        },
        {
          "Transact": {
            "call": {
              "encoded": "0x17000101010101010101010101010101010101010101010101010101010101010101",
            },
            "originKind": "Xcm",
            "requireWeightAtMost": {
              "proofSize": 10000,
              "refTime": 1000000000,
            },
          },
        },
      ],
      "messageId": "(redacted)",
      "origin": {
        "interior": {
          "X1": [
            {
              "Plurality": {
                "id": "Technical",
                "part": "Voice",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "Sent",
    "section": "polkadotXcm",
  },
]
`;
